subprojects/model-core/src/main/java/org/gradle/internal/instantiation/generator/AsmBackedClassGenerator.java

```diff
@@ -1178,6 +1181,20 @@ private BytecodeFragment getInjectedServiceGetterEpilogue(Type serviceType, Stri
             }
         }
 
+        private boolean isServiceTypeDisallowedAtExecution(Type serviceType) {
+            Class<?> serviceClass;
+
+            try {
+                serviceClass = Class.forName(serviceType.getClassName());
+            } catch (ClassNotFoundException exception) {
+                return false;
+            }
+
+            return DISALLOWED_AT_EXECUTION_INJECTED_SERVICES_CLASSES
```

Comment:
💡 Perhaps we should simply include the full set of types (the full hierarchy we want to disallow) so the implementation can go back to a quick and direct `contains` call.